nanochat/worker.py

```
@@ -0,0 +1,671 @@
+    return DecodeOutput(text=tokenizer.decode(inp.tokens)).model_dump()
+
+
+async def fn_tools_execute(data: ExecuteCodeInput) -> ExecuteCodeOutput:
+    inp = ExecuteCodeInput.model_validate(data) if isinstance(data, dict) else data
+
+    stdout_buf = io.StringIO()
+    stderr_buf = io.StringIO()
+
+    try:
+        with contextlib.redirect_stdout(stdout_buf), contextlib.redirect_stderr(stderr_buf):
+            exec(inp.code, {"__builtins__": __builtins__}, {})
```

Ruff (0.15.7)

[error] Use of `exec` detected

(S102)